cid-decode/src/main.rs

```diff
@@ -53,51 +64,151 @@ fn main() -> Result<()> {
         let entry = results.entry(res.clone()).or_default();
         *entry += 1;
 
-        s.clear();
+        buffer.clear();
     }
     // Restore panic hook.
-    panic::set_hook(h);
+    panic::set_hook(default_panic_hook);
 
     // We only get here if reading from Stdin fails...
-    rdr.read_line(&mut s)?;
+    rdr.read_line(&mut buffer)?;
 
     Ok(())
 }
 
 #[derive(Debug, Clone)]
 struct Metadata {
-    base: multibase::Base,
+    base: cid::multibase::Base,
     version: cid::Version,
-    codec: cid::Codec,
-    hash: Option<multihash::Code>,
+    codec: self::codec::Codec,
+    hash: Option<cid::multihash::Code>,
     hash_len: usize,
 }
 
-fn do_single(s: &str) -> Result<Metadata> {
-    let c = cid::Cid::try_from(s)?;
-    if c.version() == cid::Version::V0 {
-        return Ok(Metadata {
-            base: multibase::Base::Base58Btc,
-            version: c.version(),
-            codec: c.codec(),
-            hash: match std::panic::catch_unwind(|| c.hash().algorithm()) {
-                Ok(h) => Some(h),
-                Err(_) => None,
-            },
-            hash_len: c.hash().digest().len(),
-        });
-    }
-
-    let (b, _) = multibase::decode(s)?;
-
-    Ok(Metadata {
-        base: b,
+fn do_single(line: &str, line_no: usize) -> Result<Metadata> {
+    let c = cid::Cid::try_from(line).context(format!(
+        "could not parse to cid (row {}: {})",
+        line_no, line
+    ))?;
+    return Ok(Metadata {
+        base: if c.version() == cid::Version::V0 {
+            cid::multibase::Base::Base58Btc
+        } else {
+            cid::multibase::decode(line)
+                .context(format!(
+                    "could not decode multibase (row {}: {})",
+                    line_no, line
+                ))?
+                .0
+        },
         version: c.version(),
-        codec: c.codec(),
-        hash: match std::panic::catch_unwind(|| c.hash().algorithm()) {
-            Ok(h) => Some(h),
+        codec: codec::Codec::try_from(c.codec()).context(format!(
+            "could not find humanreadable name for multicodec (codec: {}, row {}: {})",
+            c.codec(),
+            line_no,
+            line
+        ))?,
+        hash: match std::panic::catch_unwind(|| c.hash().code()) {
+            Ok(h) => Some(cid::multihash::Code::try_from(h).unwrap()),
```

Comment:
These two lines are the problematic ones:

Calling `c.hash().code()` returns an integer code from a `MultihashGeneric<S>` struct, which was created when we parse the CID in line 88, using [Cid::read_bytes](https://docs.rs/cid/0.8.5/src/cid/cid.rs.html#127-141) (through a chain of conversions).
The `MultihashGeneric<S>` is created using the [read_multihash function](https://docs.rs/multihash/0.16.2/src/multihash/multihash.rs.html#298-312), which happily takes any byte sequence that looks like a multihash. Especially, it takes any _code_.

We then call `multihash::Code::try_from(<code>)` using the code parsed above (which is unchecked, verbose, from the original CID). The code for this is, I believe, generated using procedural macros, see [here](https://github.com/multiformats/rust-multihash/blob/master/derive/src/multihash.rs#L249-L259) and [here](https://docs.rs/multihash/0.16.2/src/multihash/multihash_impl.rs.html#11). Ultimately, this will fail (and thus panic) for any code that is not supported by the `multihash` crate.

And this is the point I wanted to make: We restrict ourselves to hash functions supported by the `multihash` crate, which is unnecessary, since we don't want to _use_ those functions. We only want to know which one was used to create the CID.

So what we need is a dictionary to look up the _names_  of hash functions by their multicodec code. We don't need to implement any parsing ourselves -- we reuse the parsing done by the `multihash` crate. We then take the code and size it reports, look up the name for the code, and that should be it.

Reply:
Ok I kind of get what you mean. 
I do not understand why we should encounter CID with multihash codes which are not supported by the `multihash` create. 
Since the `cid` create uses the `multihash` create these CID are then also not supported by the `cid` create. 

If we want to support more CID then the `cid` create then we shouldn't use the `cid` create to begin with. Because we have a different scope then the `cid` create